doc/sphinx/language/core/primitive.rst

```diff
@@ -150,7 +150,17 @@ extraction. Instead, it has to be provided by the user (if they want to compile
 or execute the extracted code). For instance, an implementation of this module
 can be taken from the kernel of Coq (see ``kernel/parray.ml``).
 
-Primitive arrays expose a functional interface, but they are internally
-implemented using a persistent data structure :cite:`ConchonFilliatre07wml`.
-Update and access to an element in the most recent copy of an array are
-constant time operations.
+Semantically, the set array operation behaves as if the array was first entirely
+copied, then the copy is modified and finally returned. Operationally however,
+the array copies are avoided by a mechanism loosely related to the decorator
+design pattern. The full details of the implementation are in section 2.3 of
+:cite:`ConchonFilliatre07wml`. In short, multiple versions of the array coexist
+in memory. One of those versions is special and directly stored as an OCaml
+native array. Accesses to that special version are constant time operations.
+The other versions of the array are represented as lists of modifications to
+be applied to that native version. Therefore, a version where all cells of the
+array are modified has the same access time as a list, which defeats the
+purpose of Coq's native arrays. Which version is native changes dynamically at
+each get call: the list of modifications is executed inside the OCaml native array,
+and the lists of modifications of the other versions are updated so that they
+still represent the same values.
```

Comment:
```suggestion
:cite:`ConchonFilliatre07wml`. The algorithm keeps one version as an OCaml native
array and other versions as lists of modifications to that array.
Accesses to the native array version are constant time operations.
Accesses to versions where all the cells of the array are modified has O(n) access time,
the same as a list, which defeats the purpose of Coq's native arrays. The version
that is kept as the native array can change dynamically upon
each get call: a list of modifications may be applied to the native array
and the lists of modifications of the other versions are updated so that they
still represent the same values.
```

How about something like this?  I guessed a little about your meaning, hope
I didn't distort it.

Reply:
I appreciate that saying "may" and "can" is more flexible and allows us to change the implementation in the future. However this is a feature, not an implementation detail. The re-rooting of the array by the get operation allows algorithms to precisely control their speed of calculation. The documentation should commit to that. I agree with your other modifications.

Reply:
Are you saying that every time you do a set operation, the native array and the lists of modifications are updated?  Is updating the lists still O(1)?  (FWIW, I would expect some LRU behavior for the various versions though that might not be O(1).)

By all means, make the text accurate.

Reply:
I updated my suggestion to say "have O(n) access time".

Reply:
As written in section 2.3.2 of the article, both get and set re-root the array so they are not always O(1). They are idempotent though, from the second call onwards they become O(1) if evaluated on the same version.